src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_opt.cpp

```diff
@@ -0,0 +1,1126 @@
+    return ptr->get_onednn_memory(dnnl::memory::desc(dnnl::memory::dims(dim), convert_data_type(ptr->get_layout().data_type), tag), offset);
+}
+
+class MOEOptImpl : public PrimitiveImplOCL {
+public:
+    DECLARE_OBJECT_TYPE_SERIALIZATION(ov::intel_gpu::ocl::MOEOptImpl)
+    Stage::Ptr softmax_topk = make_stage<MOEOptSoftMaxTopK>();
+    Stage::Ptr gather = make_stage<MOEOptGather>();
+    Stage::Ptr scatter = make_stage<MOEOptScatter>();
+    Stage::Ptr mlp_gate_up = make_stage<MOEOptMLPGateUp>();
+    Stage::Ptr mlp_down = make_stage<MOEOptMLPDown>();
+    Stage::Ptr mlp_reduce = make_stage<MOEOptMLPReduce>();
+
+    struct dnnl_weights {
+        dnnl::memory weight;
+        dnnl::memory scale;
+        dnnl::memory zp;
+        int ic, oc, ic_group_size;
+    };
+
+    // expert_mask result in cpu side
+    struct expert_mask_cpu {
+        std::vector<int8_t> pred_flag;
+        // shape: [expert_num, batch_no]
+        std::vector<std::vector<int>> batch;
+        // shape: [expert_num, topk_no]
+        std::vector<std::vector<int>> topk;
+    };
+
+    // store expert_mask for gpu kernel
+    struct expert_mask_gpu {
+        memory::ptr batch;
+        memory::ptr topk;
+    };
+
+    struct moe_fusion_weights_base_addr {
+        memory::ptr weight[3];  // gate/up/down weights, experts fusion
+        memory::ptr scale[3];
+        memory::ptr zp[3];
+        memory::ptr bias[3];
+    } moe_fusion_weights;
+
+    struct scratch_buffers {
+        // softmax+topk
+        memory::ptr topk_id;
+        memory::ptr topk_weights;
+
+        // fast single batch: scratch.up = up(x) * silu(gate(x))
+        //                    scratch.y = down(scratch.up) * routing_weights
+        memory::ptr up;
+        memory::ptr y;
+        // onednn: scratch.x, scratch.routing_weights = gather(x, ...)
+        //         scratch.up = up(scratch.x)
+        //         scratch.gate = gate(scratch.x) * scratch.up
+        //         scratch.y = down(scratch.gate) * routing_weights
+        memory::ptr x;
+        memory::ptr routing_weights;
+        memory::ptr gate;
+        // buffers for batch and topk from cpu, each expert has one
+        std::vector<expert_mask_gpu> expert_masks;
+
+        moe_fusion_weights_base_addr moe_fusion_wei_addr;
+        memory::ptr input_routing_weights;
+        memory::ptr input_router_topk_idx;
+    };
+
+    std::vector<std::vector<dnnl_weights>> _dnnl_weights;
+    int _hidden_size;
+    int _intermediate_size;
+    int _group_size;
+
+    MOEOptImpl() : PrimitiveImplOCL(MOEOpt::get_type_info_static()) {}
+    MOEOptImpl(const program_node& node, const RuntimeParams& params) : MOEOptImpl() {
+        init(node.as<moe_fused_compressed>().get_primitive());
+
+        add_stage(softmax_topk, params);
+        add_stage(gather, params);
+        add_stage(scatter, params);
+        add_stage(mlp_gate_up, params);
+        add_stage(mlp_down, params);
+        add_stage(mlp_reduce, params);
```

Comment:
I think we can 
1) writhe subkernels in separate files 
2) use each kernel or not by controllling the stage?